File: extract-config/extract-config.go
// extract-config — pulls the embedded Psiphon network config out of an
// official Conduit release binary and writes it as psiphon_config.json
// in the current directory.
//
// The official Conduit binary (from Psiphon's GitHub Releases) embeds its
// psiphon_config.json via Go's //go:embed directive. This tool finds that
// JSON blob inside the binary and copies it to a file. It's only copying
// data that's already in the binary you downloaded — nothing fancy.
//
// Usage:
//   extract-config <path-to-official-conduit-binary>
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
)

const outputFile = "psiphon_config.json"

func main() {
	if len(os.Args) != 2 {
		fmt.Fprintln(os.Stderr, "Usage: extract-config <path-to-official-conduit-binary>")
		fmt.Fprintln(os.Stderr, "")
		fmt.Fprintln(os.Stderr, "Tip: just type './extract-config ' then drag the official")
		fmt.Fprintln(os.Stderr, "Conduit binary into your terminal window. It will paste the path.")
		os.Exit(2)
	}

	data, err := os.ReadFile(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot read file: %v\n", err)
		os.Exit(1)
	}

	needle := []byte(`"PropagationChannelId"`)
	pos := bytes.Index(data, needle)
	if pos < 0 {
		fmt.Fprintln(os.Stderr, "Could not find a Psiphon config inside this file.")
		fmt.Fprintln(os.Stderr, "Are you sure this is an official Conduit binary from")
		fmt.Fprintln(os.Stderr, "https://github.com/Psiphon-Inc/conduit/releases ?")
		os.Exit(1)
	}

	start := bytes.LastIndexByte(data[:pos], '{')
	if start < 0 {
		fmt.Fprintln(os.Stderr, "Found marker but no opening brace before it.")
		os.Exit(1)
	}

	end := -1
	depth, inStr, esc := 0, false, false
	for i := start; i < len(data); i++ {
		c := data[i]
		if esc {
			esc = false
			continue
		}
		if c == '\\' && inStr {
			esc = true
			continue
		}
		if c == '"' {
			inStr = !inStr
			continue
		}
		if inStr {
			continue
		}
		switch c {
		case '{':
			depth++
		case '}':
			depth--
			if depth == 0 {
				end = i + 1
				break
			}
		}
		if end >= 0 {
			break
		}
	}
	if end < 0 {
		fmt.Fprintln(os.Stderr, "Could not find end of JSON object.")
		os.Exit(1)
	}

	blob := data[start:end]
	if err := json.Unmarshal(blob, new(any)); err != nil {
		fmt.Fprintf(os.Stderr, "Extracted bytes don't parse as JSON: %v\n", err)
		os.Exit(1)
	}

	if err := os.WriteFile(outputFile, blob, 0o644); err != nil {
		fmt.Fprintf(os.Stderr, "Could not write %s: %v\n", outputFile, err)
		os.Exit(1)
	}

	fmt.Fprintf(os.Stderr, "✓ Wrote %s (%d bytes) in the current folder.\n", outputFile, len(blob))
}

File: extract-config.py
#!/usr/bin/env python3
"""
extract-config.py — pull the embedded Psiphon network config out of an
official Conduit release binary.

The official Conduit binary (downloadable from
https://github.com/Psiphon-Inc/conduit/releases) embeds its psiphon_config.json
via Go's //go:embed directive. This script finds that JSON blob inside the
binary and writes it to a file. This is just copying data that's already
in the binary you downloaded — it's not bypassing anything.

Usage:
    python3 extract-config.py /path/to/official/conduit-binary > psiphon_config.json

Works on macOS, Linux, and Windows. Requires Python 3 (already installed on
macOS and most Linux distros).
"""

import json
import re
import sys


def extract(path: str) -> bytes:
    with open(path, "rb") as f:
        data = f.read()

    needle = b'"PropagationChannelId"'
    pos = data.find(needle)
    if pos == -1:
        raise SystemExit(
            f"Could not find Psiphon config inside {path}. "
            "Are you sure this is an official Conduit binary?"
        )

    # Walk backward to the opening brace of the JSON object.
    start = data.rfind(b"{", 0, pos)
    if start == -1:
        raise SystemExit("Found marker but no opening brace before it.")

    # Walk forward, balancing braces and respecting string boundaries.
    depth = 0
    in_str = False
    esc = False
    end = None
    for i in range(start, len(data)):
        c = data[i : i + 1]
        if esc:
            esc = False
            continue
        if c == b"\\" and in_str:
            esc = True
            continue
        if c == b'"':
            in_str = not in_str
            continue
        if in_str:
            continue
        if c == b"{":
            depth += 1
        elif c == b"}":
            depth -= 1
            if depth == 0:
                end = i + 1
                break

    if end is None:
        raise SystemExit("Could not find end of JSON object.")

    blob = data[start:end]
    # Validate it parses.
    try:
        json.loads(blob)
    except json.JSONDecodeError as exc:
        raise SystemExit(f"Extracted bytes don't parse as JSON: {exc}")

    return blob


def main() -> None:
    if len(sys.argv) != 2 or sys.argv[1] in ("-h", "--help"):
        print(__doc__.strip(), file=sys.stderr)
        sys.exit(1)
    blob = extract(sys.argv[1])
    sys.stdout.buffer.write(blob)
    sys.stdout.buffer.write(b"\n")


if __name__ == "__main__":
    main()
